src/modules/ellsq/main_impl.h

```diff
@@ -0,0 +1,305 @@
+        uint32_t cnt = 0;
+        /* Field elements and branch values are extracted from
+         * SHA256("secp256k1_ellsq_encode\x00" + uint32{cnt} + rnd32 + X + byte{Y & 1})
+         * for consecutive values of cnt. cnt==0 is first used to populate a pool of
+         * 128 2-bit branch values. The 128 cnt values that follow are used to
+         * generate field elements u1. cnt==129 (and multiples thereof) are used to
+         * repopulate the pool and start over, if that were ever necessary. */
+        unsigned char hashdata[23 + 4 + 32 + 32 + 1] = "secp256k1_ellsq_encode";
+        /* Pool of 2-bit branch values. */
+        unsigned char branch_hash[32];
+        /* Number of 2-bit values in branch_hash left. */
+        int branches_left = 0;
+        /* Fill up hashdata, excluding i. */
+        memcpy(hashdata + 23 + 4, rnd32, 32);
+        secp256k1_fe_get_b32(hashdata + 23 + 4 + 32, &p.x);
+        hashdata[4 + 23 + 32 + 32] = secp256k1_fe_is_odd(&p.y);
+        while (1) {
+            int branch;
+            secp256k1_fe u1, u2;
+            secp256k1_ge q;
+            secp256k1_gej qj;
+            /* If the pool of branch values is empty, populate it. */
+            if (branches_left == 0) {
+                secp256k1_sha256 hash;
+                hashdata[23 + 0] = cnt;
+                hashdata[23 + 1] = cnt >> 8;
+                hashdata[23 + 2] = cnt >> 16;
+                hashdata[23 + 3] = cnt >> 24;
+                secp256k1_sha256_initialize(&hash);
+                secp256k1_sha256_write(&hash, hashdata, sizeof(hashdata));
+                secp256k1_sha256_finalize(&hash, branch_hash);
+                ++cnt;
+                branches_left = 128;
+            }
+            /* Take a 2-bit branch value from the branch pool. */
+            --branches_left;
+            branch = (branch_hash[(127 - branches_left) >> 2] >> (((127 - branches_left) & 3) << 1)) & 3;
+            /* Compute a new u1 value by hashing (a potential first 32 bytes of the output). */
+            {
+                secp256k1_sha256 hash;
+                hashdata[23 + 0] = cnt;
+                hashdata[23 + 1] = cnt >> 8;
+                hashdata[23 + 2] = cnt >> 16;
+                hashdata[23 + 3] = cnt >> 24;
+                secp256k1_sha256_initialize(&hash);
+                secp256k1_sha256_write(&hash, hashdata, sizeof(hashdata));
+                secp256k1_sha256_finalize(&hash, ell64);
+                ++cnt;
+            }
+            if (!secp256k1_fe_set_b32(&u1, ell64)) continue;
+            /* Compute the remainder Q to encode in the last 32 bytes of the output. */
+            secp256k1_ellsq_fe_to_ge_var(&q, &u1);
+            secp256k1_ge_neg(&q, &q);
+            secp256k1_gej_set_ge(&qj, &q);
+            secp256k1_gej_add_ge_var(&qj, &qj, &p, NULL);
+            if (!secp256k1_gej_is_infinity(&qj)) {
+                 secp256k1_ge_set_gej_var(&q, &qj);
+            } else {
```

Comment:
Quick question, is there an easy way to hit this branch in a test, that doesn't involve something convoluted like mocking out the hash function?

No worries if not but just thought I'd double check, since I added a `VERIFY_CHECK(0)` here and it doesn't get hit by either the test or benchmark.

Reply:
See the comment below:

>  Note that if u1 is generated using a secure hash function, the probability of reaching this branch is negligible.

So I think it's indeed impossible to construct a testcase for it without stubbing out the hash function.